src/utils/errors.ts

```diff
@@ -6,7 +6,109 @@ export const isErrorNodeSystemError = (error: unknown): error is NodeJS.ErrnoExc
     'code' in (error as never) &&
-    'path' in (error as never) &&
     'syscall' in (error as never)
   )
 }
+
+export enum ErrorCode {
+  GENERIC = 1000,
+  COMMAND_UNKNOWN,
+  API_KEY_MISSING,
+  API_KEY_INVALID,
+  API_RESPONSE_400,
+  API_RESPONSE_429,
+  API_RESPONSE_500,
+  LEAGUE_NOT_FOUND_BY_NAME,
+  LEAGUE_NOT_FOUND_BY_CODE,
+  TEAM_NOT_FOUND,
+  CACHE_WRITE,
+  NETWORK_UNREACHABLE,
+}
+
+export class ApplicationError extends Error {
+  constructor(code: ErrorCode.GENERIC, message: string)
+  constructor(code: ErrorCode.COMMAND_UNKNOWN, command: string)
+  constructor(code: ErrorCode.API_KEY_MISSING)
+  constructor(code: ErrorCode.API_KEY_INVALID)
+  constructor(code: ErrorCode.API_RESPONSE_400, message: string)
+  constructor(code: ErrorCode.API_RESPONSE_429)
+  constructor(code: ErrorCode.API_RESPONSE_500)
+  constructor(code: ErrorCode.TEAM_NOT_FOUND, teamName: string)
+  constructor(code: ErrorCode.LEAGUE_NOT_FOUND_BY_NAME, leagueName: string)
+  constructor(code: ErrorCode.LEAGUE_NOT_FOUND_BY_CODE, leagueCode: string)
+  constructor(code: ErrorCode.CACHE_WRITE)
+  constructor(code: ErrorCode.NETWORK_UNREACHABLE)
+  constructor(public code: ErrorCode, public extraData?: string) {
+    super()
+  }
+}
+
+export const formatErrorForPrinting = (code: ErrorCode, extraData?: string): string => {
+  switch (code) {
+    case ErrorCode.GENERIC:
+      return `Something unexpected happened:\n     ${extraData}\n`
+
+    case ErrorCode.COMMAND_UNKNOWN:
+      return `Unknown command "${extraData}".\n`
+
+    case ErrorCode.API_KEY_MISSING:
+      return (
+        'SOCCER_GO_API_KEY environment variable not set.\n\n' +
+        '    $ export SOCCER_GO_API_KEY=<football_data_api_key>\n\n' +
+        'You can get your own API key over at\n' +
+        'https://www.football-data.org/client/register\n'
+      )
+
+    case ErrorCode.API_KEY_INVALID:
+      return (
+        'The API key set is invalid.\n' +
+        'Please check that the SOCCER_GO_API_KEY environment variable matches ' +
+        'the key you received upon registration\n\n' +
+        '    $ echo $SOCCER_GO_API_KEY\n'
+      )
+
+    case ErrorCode.API_RESPONSE_400:
+      return `Something went wrong with your request, reason:\n\n    ${extraData}\n`
+
+    case ErrorCode.API_RESPONSE_429:
+      return (
+        'You seem to have made too many requests.\n' +
+        'Please wait a minute before making new requests.\n'
+      )
+
+    case ErrorCode.API_RESPONSE_500:
+      return 'Cannot retrieve data at this time, please try again later.\n'
+
+    case ErrorCode.LEAGUE_NOT_FOUND_BY_NAME:
+      return `Could not find league "${extraData}".\n`
+
+    case ErrorCode.LEAGUE_NOT_FOUND_BY_CODE:
+      return `Could not find league with code "${extraData}".\n`
+
+    case ErrorCode.TEAM_NOT_FOUND:
+      return `Could not find team "${extraData}".\n`
+
+    case ErrorCode.CACHE_WRITE:
+      return `There was an issue when attempting to write data to cache.\n`
```

Comment:
IMHO this error should not be fatal, as in: the cache is entirely optional and if anything regarding the cache fails we should ignore it (or print a warning - but not quit the program)

Reply:
Makes sense. I kept it as fatal because that was the behavior I found before working on this (see https://github.com/acifani/soccer-go/pull/396/files#diff-f0b1423191726c0f2981ced42942c5c7743094bb86c9c05495ca9fa00a855583L22).

I'll update the PR to not fail in this case.